server/serverimpl_test.go

```diff
@@ -263,10 +281,77 @@ func TestDisconnectWSConnection(t *testing.T) {
 	assert.NoError(t, err)
 	assert.True(t, atomic.LoadInt32(&connectionCloseCalled) == 0)
 
-	// Close connection from server side
-	srvConn := wsConnection{wsConn: conn}
-	err = srvConn.Disconnect()
+	// Close connection from client side
+	clientConn := wsConnection{wsConn: conn, connMutex: &sync.Mutex{}, closed: &atomic.Bool{}}
+	err = clientConn.Disconnect()
+	assert.NoError(t, err)
+
+	// Verify connection disconnected from server side
+	eventually(t, func() bool { return atomic.LoadInt32(&connectionCloseCalled) == 1 })
+	// Waiting for wsConnection to fail ReadMessage() over a Disconnected communication
+	eventually(t, func() bool {
+		_, _, err := conn.ReadMessage()
+		return err != nil
+	})
+}
+
+// testLogger is a struct that adapts a *zap.Logger to opamp-go's Logger interface.
+type testLogger struct {
+	logs []string
+}
+
+func newTestLogger() *testLogger {
+	return &testLogger{
+		logs: []string{},
+	}
+}
+
+func (o *testLogger) Debugf(_ context.Context, format string, v ...any) {
+	log := fmt.Sprintf(format, v...)
+	o.logs = append(o.logs, fmt.Sprintf("Debugf: %s\n", log))
+}
+
+func (o *testLogger) Errorf(_ context.Context, format string, v ...any) {
+	log := fmt.Sprintf(format, v...)
+	o.logs = append(o.logs, fmt.Sprintf("Errorf: %s\n", log))
+}
+
+func TestDisconnectServerWSConnection(t *testing.T) {
+	connectionCloseCalled := int32(0)
+	var serverConn types.Connection
+	mutex := sync.Mutex{}
+	callback := types.Callbacks{
+		OnConnecting: func(request *http.Request) types.ConnectionResponse {
+			return types.ConnectionResponse{Accept: true, ConnectionCallbacks: types.ConnectionCallbacks{
+				OnConnected: func(ctx context.Context, conn types.Connection) {
+					mutex.Lock()
+					serverConn = conn
+					mutex.Unlock()
+				},
+				OnConnectionClose: func(conn types.Connection) {
+					atomic.StoreInt32(&connectionCloseCalled, 1)
+				},
+			}}
+		},
+	}
+
+	// Start a Server.
+	logger := newTestLogger()
+	settings := &StartSettings{Settings: Settings{Callbacks: callback}}
+	srv := startServerWithLogger(t, settings, logger)
+	defer srv.Stop(context.Background())
+
+	// Connect to the Server.
+	conn, _, err := dialClient(settings)
+
+	// Verify that the connection is successful.
 	assert.NoError(t, err)
+	assert.True(t, atomic.LoadInt32(&connectionCloseCalled) == 0)
+
+	// Close connection from server side
+	mutex.Lock()
```

Comment:
What does this mutex protect against? Does `serverConn` value ever change (except from `nil` to the only connection we get)? Can `serverConn` be nil here?